nixos/modules/services/web-apps/bookstack.nix

```diff
@@ -0,0 +1,319 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.bookstack;
+  bookstack = cfg.package;
+  db = cfg.database;
+  mail = cfg.mail;
+
+  user = "bookstack";
+  group = config.services.nginx.group;
+
+  phpEnv = {
+    APP_KEY = "\"base64:${cfg.appKey}\"";
+    APP_URL = cfg.appURL;
+    DB_HOST = db.host;
+    DB_PORT = toString db.port;
+    DB_DATABASE = db.name;
+    DB_USERNAME = db.user;
+    MAIL_DRIVER = mail.driver;
+    MAIL_FROM_NAME = "\"${mail.fromName}\"";
+    MAIL_FROM = mail.from;
+    MAIL_HOST = mail.host;
+    MAIL_PORT = toString mail.port;
+  } // optionalAttrs (db.passwordFile != null) {
+    DB_PASSWORD = "\"$(head -n1 ${db.passwordFile})\"";
+  } // optionalAttrs (mail.user != null) {
+    MAIL_USERNAME = mail.user;
+  } // optionalAttrs (mail.passwordFile != null) {
+    MAIL_PASSWORD = "\"$(head -n1 ${mail.passwordFile})\"";
+  } // optionalAttrs (mail.encryption != null) {
+    MAIL_ENCRYPTION = mail.encryption;
+  };
+
+in {
+  options.services.bookstack = {
+
+    enable = mkEnableOption ''
+      Enable BookStack.
+      The default admin details are "admin@admin.com" with a password of "password". You should change these details immediately after logging in for the first time.
+      This will also enable <option>services.phpfpm</option> and <option>services.nginx</option>.
+    '';
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.bookstack.override {
+        dataDir = cfg.dataDir;
+      };
+      defaultText = "pkgs.bookstack";
+      description = "BookStack package to use.";
+    };
+
+    appKey = mkOption {
```

Comment:
Not suitable for storing in the nix store.

Reply:
I'm sorry, I'm not quite sure what you're saying. I've basically copied this way of working from some other packages (I think DokuWiki was one of them).

Reply:
`appKey` is an encryption key - a secret. If you do this the `appKey` will show up under `/nix/store` in a file that is readable by anyone on the system. You should not be storing encryption keys world readable like this, it promotes "bad" behavior. Maybe DokuWiki needs a PR to fix that.